src/custom_request_schedule_manager.cc

```diff
@@ -0,0 +1,121 @@
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "custom_request_schedule_manager.h"
+
+namespace triton::perfanalyzer {
+
+cb::Error
+CustomRequestScheduleManager::Create(
+    const pa::PAParamsPtr& params, const std::shared_ptr<ModelParser>& parser,
+    const std::shared_ptr<cb::ClientBackendFactory>& factory,
+    std::unique_ptr<LoadManager>* manager)
+{
+  std::unique_ptr<CustomRequestScheduleManager> local_manager(
+      new CustomRequestScheduleManager(params, parser, factory));
+
+  *manager = std::move(local_manager);
+
+  return cb::Error::Success;
+}
+
+CustomRequestScheduleManager::CustomRequestScheduleManager(
+    const pa::PAParamsPtr& params, const std::shared_ptr<ModelParser>& parser,
+    const std::shared_ptr<cb::ClientBackendFactory>& factory)
+    : RequestRateManager(
+          params->async, params->streaming, Distribution::CUSTOM,
+          params->batch_size, params->measurement_window_ms, params->max_trials,
+          params->max_threads, params->num_of_sequences,
+          params->shared_memory_type, params->output_shm_size,
+          params->serial_sequences, parser, factory,
+          params->request_parameters),
+      schedule_(params->schedule)
+{
+}
+
+cb::Error
+CustomRequestScheduleManager::PerformWarmup(
+    double request_rate, size_t warmup_request_count)
+{
+  if (warmup_request_count == 0) {
+    return cb::Error::Success;
+  }
+  RETURN_IF_ERROR(ChangeRequestRate(request_rate, warmup_request_count));
+  WaitForWarmupAndCleanup();
+  return cb::Error::Success;
+}
+
+cb::Error
+CustomRequestScheduleManager::ChangeRequestRate(
+    const double request_rate, const size_t request_count)
+{
+  PauseWorkers();
+  ConfigureThreads(request_count);
+  GenerateSchedule(request_rate, schedule_);
+  ResumeWorkers();
+
+  return cb::Error::Success;
+}
+
+void
+CustomRequestScheduleManager::GenerateSchedule(
+    const double request_rate, const std::vector<float>& schedule)
+{
+  std::vector<float> scaled_schedule;
+  scaled_schedule.reserve(schedule.size());
+  if (schedule.size() > 0) {
+    for (const auto& value : schedule) {
+      scaled_schedule.push_back(value / static_cast<float>(request_rate));
```

Comment:
@IzzyPutterman why are the values from schedule (e.g. `--schedule=1,2,3,4,4.5,6,9,10,11,13,15,18` being divided by the request rate?

I thought those values are supposed to be absolute timestamps?

cc @lkomali @debermudez

Reply:
I added a random feature so I could scale by the request-rate if I wanted.  So without needing to change the schedule file I could speedup the sending by 2x or slowdown by 2x.

Reply:
I think this should be a new option, it kinda breaks the previously-communicated use of the `--request-rate-range` option.